batch/batch/front_end/query/query_v2.py

```diff
@@ -125,40 +125,46 @@ def parse_list_batches_query_v2(user: str, q: str, last_batch_id: Optional[int])
         where_conditions.append(f'({cond})')
         where_args += args
 
-    sql = f"""
-SELECT batches.*,
-  cancelled_t.cancelled IS NOT NULL AS cancelled,
+    sql = f"""\
+SELECT
+  batches.*,
+  cancelled_t.cancelled <=> 1 AS cancelled,
   job_groups_n_jobs_in_complete_states.n_completed,
   job_groups_n_jobs_in_complete_states.n_succeeded,
   job_groups_n_jobs_in_complete_states.n_failed,
   job_groups_n_jobs_in_complete_states.n_cancelled,
-  cost_t.cost, cost_t.cost_breakdown
-FROM job_groups
-LEFT JOIN batches ON batches.id = job_groups.batch_id
-LEFT JOIN billing_projects ON batches.billing_project = billing_projects.name
-LEFT JOIN job_groups_n_jobs_in_complete_states ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
-LEFT JOIN LATERAL (
-  SELECT 1 AS cancelled
-  FROM job_group_self_and_ancestors
-  INNER JOIN job_groups_cancelled
-    ON job_group_self_and_ancestors.batch_id = job_groups_cancelled.id AND
-      job_group_self_and_ancestors.ancestor_id = job_groups_cancelled.job_group_id
-  WHERE job_groups.batch_id = job_group_self_and_ancestors.batch_id AND
-    job_groups.job_group_id = job_group_self_and_ancestors.job_group_id
-) AS cancelled_t ON TRUE
-STRAIGHT_JOIN billing_project_users ON batches.billing_project = billing_project_users.billing_project
-LEFT JOIN LATERAL (
-  SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
-  FROM (
-    SELECT resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
+  cost_t.cost,
+  cost_t.cost_breakdown
+FROM batches
+INNER JOIN job_groups
```

Comment:
I might benefit from a in-person explanation for why these are the same. It seems possible that

`job_groups LEFT JOIN batches` would give a different result to `batches INNER JOIN job_groups`

Like... wouldn't a job group that isn't represented in batches would appear in the left query but not the right query? 

But maybe there's some data-model reason why that's not the case? (eg that it's impossible to be a job group without a batch?)

This question also applies for all the other joins down the chain

Reply:
The aim of the query is to return a list of batches annotated with some high level metadata like cost, job statuses and whether or not the batch has been cancelled.

To convince yourself that they're the same, you also need to look at the data in the where-clause (defined above this f-string). Of note, we're filtering for job groups with `job_group_id = 0`. 

All batches have at least one job group with `job_group_id = 0` (ie the root job group). Performing and inner join or left join will yield the same answer. Operationally I'm doing `batches` `inner join` (`job_groups` filtered by `job_group_id = 0`).

A job group cannot exist outside of a batch.

Reply:
A batch is cancelled when a record exists in the `job_groups_cancelled` table with the same id and job_group_id = 1, ie:
```sql
SELECT B.id
     , J.id IS NOT NULL
FROM batches B
LEFT JOIN job_groups_cancelled J
  ON B.id = J.id
WHERE
  J.job_group_id = 0
;
```

Reply:
For efficency, we do not add a record into `job_groups_cancelled` for every cancelled job gorup. Instead, a job group is cancelled if a record exists in `job_groups_cancelled` for an ancestral job group.

The table `job_group_self_and_ancestors` tells us the job group `ancestor_id`  for each job group. For the root job group, this id is itself (as the table name implies).

```sql
SELECT G.batch_id, G.job_group_id, C.job_group_id IS NOT NULL AS cancelled
FROM job_groups AS G
INNER JOIN job_groups_self_and_ancestors AS A
   ON G.batch_id     = A.batch_id
  AND G.job_group_id = A.job_group_id
LEFT JOIN job_groups_cancelled AS C
   ON G.batch_id    = C.id
  AND A.ancestor_id = C.job_group_id;
```

Reply:
As I understand it, `job_groups_n_jobs_in_complete_states` contains aggregated information for the job group hierarchy so we only need the record for job_group_id 0.

Reply:
The lateral join on the table derived from `aggregated_job_group_costs` is to avoid computing the costs of the batch for all batches and then joining.

Reply:
This all is beginning to make sense. Since I tripped over this many times understanding the rest of the query, how would you feel about doing something like `INNER JOIN job_groups AS root_job_group` here, to make it a little easier for humans to parse the tables' name/intention/usage in the rest of the query?

Reply:
That might indeed make the query more human readable, but names like `job_groups` are exposed to user defined queries that you see in other files. A follow-up change would improve that but harder to test.